### PR TITLE
feature: add --stop-on-first-error CLI option

### PR DESCRIPTION
**Pull request by Alizter**
We expose a CLI option called --stop-on-first-error that exposes to the user the internal behaviour of stopping the build on the first encountered error.

- [x] option documentation
- [x] changelog
- [x] test

### CODE REVIEW

File: doc/changes/8400.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- Add `--stop-on-first-error` option to `dune build` which will terminate the build when
+  the first error message is encountered. (#8400, @Alizter)
```
Comment (rgrinberg) [inline doc/changes/8400.md:2]: Let's mention the name of whoever implemented the feature in the engine as well (pmwhite I believe)

Comment (Alizter) [reply doc/changes/8400.md:2]: Yes, in fact I will probably remove my name completely if @pmwhite is happy with me writing that.

Comment (rgrinberg) [reply doc/changes/8400.md:2]: You did some work too, so keep it.

**Comment by Alizter** [comment]
@rgrinberg Do we want this configurable from dune config? I could add an option, but we don't have any precedent for a boolean option.

**Comment by rgrinberg** [comment]
I don't see the need. The feature seems very niche apart from some integrations with CI

File: test/blackbox-tests/test-cases/stop-on-first-error.t/run.t [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Here we demonstrate the --stop-on-first-error behaviour.
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/stop-on-first-error.t/run.t:1]: Can you change the test not to rely on OCaml? Simple rules and actions will do here.

**Comment by MisterDA** [comment]
Related issue on opam: https://github.com/ocaml/opam/issues/4629. Would be nice if both tools could use the same name for the switch. IMO Dune is a bit verbose.

**Pull request merged**